tests/conftest.py

```diff
@@ -1,10 +1,218 @@
+
+@pytest.fixture(scope="module")
+def auth_tb_loading_datasets(branch):
+    with testbook(
+        "docs/walkthrough/loading-datasets.ipynb",
+        # TODO(bjb): next iteration, try to execute=True against this nb
+        execute=False,
+    ) as tb:
+        verify_auth()
+        change_branch(tb, branch)
+        tb.inject(ensure_gcp_project())
+        yield tb
+
+
+@pytest.fixture(scope="module")
+def auth_tb_loading_models(branch):
+    with testbook(
+        "docs/walkthrough/loading-models.ipynb",
+        execute=False,
+    ) as tb:
+        verify_auth()
+        change_branch(tb, branch)
+        tb.inject(ensure_gcp_project())
+        yield tb
+
+
+@pytest.fixture(scope="module")
+def auth_tb_serving_models(branch):
+    with testbook(
+        "docs/walkthrough/serving-models.ipynb",
+        execute=False,
+    ) as tb:
+        verify_auth()
+        change_branch(tb, branch)
+        tb.inject(ensure_gcp_project())
+        yield tb
+
+
+@pytest.fixture(scope="session", autouse=True)
+def gcp_setup(auth_tb_quickstart):
+    threads, stop_event = start_watches()  # start watches in threads
+    for attempt in range(3):  # Retry up to 3 times
+        logger.debug(f"Attempt {attempt} to execute installer gcp-up")
+        try:
+            auth_tb_quickstart.execute_cell("installer gcp-up")
+            assert "Apply complete!" in auth_tb_quickstart.cell_output_text(
+                "installer gcp-up"
+            )
+            break
+        except Exception as err:
+            logger.warning(f"gcp-up encountered an error: {err}")
+            if attempt == 1:
+                delete_state_lock()
+            continue
+
+    yield  # teardown below the yield
+
+    logger.debug("Tearing down gcp_setup")
+    try:
+        for attempt in range(3):  # Retry up to 3 times
+            try:
+                auth_tb_quickstart.execute_cell("installer gcp-down")
+                assert "Destroy complete!" in auth_tb_quickstart.cell_output_text(
+                    "installer gcp-down"
+                )
+                break
+            except Exception as err:
+                logger.warning(f"gcp-down encountered an error: {err}")
+                if attempt == 1:
+                    delete_state_lock()
+                continue
+    finally:
+        try:
+            stop_event.set()
+            for thread in threads:
+                thread.join(timeout=1)
+            # Check for any threads that did not stop
+            alive_threads = [thread for thread in threads if thread.is_alive()]
+            if alive_threads:
+                logger.warning(
+                    f"Threads {', '.join(thread.name for thread in alive_threads)} did not stop as expected"
+                )
+
+                # Determine the exit code based on test statuses
+                exit_code = (
+                    1 if any(not status for status in test_statuses.values()) else 0
+                )
+                sys.exit(exit_code)
```

Comment:
The threads that emit logs don't seem able to be terminated gracefully. We need to shut them down with force by killing the parent process (pytest itself) and doing it with the exit code derived from the test results